### PR TITLE
Fix race condition in agent cache

### PR DESCRIPTION
**Pull request by azdagron**
As reported by @loveyana in #2696, there is a race condition in the agent cache when reading/writing the selector index. This is caused by the getSelectorIndex function opportunistically populating the index when queried for selectors. This function is however called by goroutines under both read and write locks. If a reader calls this function with a selector that does not exist in the index, then the map is mutated, possibly while being concurrently accessed by other readers, causing the race.

The fix is to have two functions for accessing the selector index with one for reads and one for writes. The write function has the current behavior and will mutate the map when needed to insert a selector into the index. The read function will return a nil index for selectors that don't exist into the index. This prevents the map from being mutated during reads.

### CODE REVIEW

File: pkg/agent/manager/cache/cache.go [hunk 1]
```diff
@@ -693,6 +695,15 @@ func (c *Cache) getSelectorIndex(s selector) *selectorIndex {
 	return index
 }
 
+// getSelectorIndexForRead gets the selector index for the selector. If one
+// doesn't exist, nil is returned.
+func (c *Cache) getSelectorIndexForRead(s selector) *selectorIndex {
+	if index, ok := c.selectors[s]; ok {
```
Comment (loveyana) [inline pkg/agent/manager/cache/cache.go:701]: We changed the expected behavior from adding a key to the `c.selectors` to directly returning `newSelectorIndex`. Is this in line with expectations, and is there an exception due to no update `c.selectors`?

**Comment by amoore877** [comment]
Just carrying forward thread from previous PR, https://github.com/spiffe/spire/pull/2696/files#r793754784

Want to confirm adding unit testing for races should wait for #2379 and be a separately filed issue

**Review by zmt — COMMENTED** [review]
Should we warn that callers must hold the read lock here similar to the warning added to `getSelectorIndexForWrite`?

**Comment by azdagron** [comment]
> Want to confirm adding unit testing for races should wait for #2379 and be a separately filed issue

That'd be my vote :)

> Should we warn that callers must hold the read lock here similar to the warning added to getSelectorIndexForWrite?
Sure, I'll fix up the comment.

**Review by rturner3 — APPROVED** [review]
LGTM

**Pull request merged**